### ISSUE TITLE
Regression in include functionality

### ISSUE THREAD

**Issue by CamJN** [microsoft/pyright#9039]
**Describe the bug**
A recent update to pyright caused it to stop being able to find code in a previously-found location.

I have the following pyright.json:
```json
{
  "include": [
    "src",
    "*.py"
  ],

  "exclude": [
    "**/__pycache__"
  ],

  "defineConstant": {
    "DEBUG": true
  },

  "reportMissingImports": true,
  "reportMissingTypeStubs": false,
  "typeCheckingMode": "strict",
  "reportPrivateUsage": false,

  "pythonVersion": "3.11",
  "venvPath": "../",
  "venv": "llm-api"
}
```

Previously pyright could find code in the src dir, now it cannot. The only change being an update to pyright.

**Code or Screenshots**
Imports of anything in the `src` dir fail.
For example:
```python
from uuids import uuid
```
Results in this error:
`Pyright [reportMissingImports]: Import "uuids" could not be resolved`

**VS Code extension or command-line**
CLI tool `pyright 1.1.381`
Run as: `/usr/local/bin/pyright-langserver --stdio`

**Comment by erictraut** [comment]
I can't think of any recent changes in pyright that would impact this functionality.

I'll need more details from you to repro the issue. Can you provide detailed repro steps or a sample project that exhibits the problem?

Also, can you confirm that an older version of pyright does not exhibit the same behavior as 1.1.381? If there was a change in behavior, in which version did it start?

**Comment by CamJN** [comment]
Yes, I can confirm that many older versions of pyright do not exhibit the same behaviour as 1.1.381, that's what I was trying to convey in my report. It used to work, now it doesn't.

In which version did it start?
Not sure, I'll have to try to reinstall some older versions, which homebrew makes an absolute pain. I'll get back to this.

Detailed repro steps and sample project that exhibits the problem:
```bash
mkdir -p ~/Developer/Python/pyright_bug_demo/src
cd ~/Developer/Python
python3.11 -m venv pyright_bug_demo
# copy all files from https://gist.github.com/CamJN/56bebd08b1949c0d2ee359c6b01ffc6c into ~/Developer/Python/pyright_bug_demo
cd ~/Developer/Python/pyright_bug_demo/
mv src_utils.py src/utils.py
source bin/activate
pyright -p .
```

**Comment by erictraut** [comment]
I'm a bit confused by those instructions. They appear to call for the project source files to be placed in the venv directory. I suspect that was unintended, so I modified the instructions to place the venv outside of the project directory.

```
repro
  - pyright_bug_demo # this is the venv
  - project
     - app.py
     - pyrightconfig.json
     - src
       - utils.py
```

Does that look right?

When I do this and then run the pyright CLI from the project directory, I'm not able to repro the behavior you're seeing. Instead, I see this output, which is what I'd expect:

```
/Users/eric/Desktop/repro/project/app.py
  /Users/eric/Desktop/repro/project/app.py:1:19 - error: Import "uuid" is not accessed (reportUnusedImport)
1 error, 0 warnings, 0 informations 
```

Incidentally, I noticed that you're using the "venv" and "venvPath" configuration approach. This is not recommended. You will typically see much better results if you simply activate the target venv prior to invoking pyright. See [this documentation](https://microsoft.github.io/pyright/#/import-resolution?id=configuring-your-python-environment) for details.

I'll also note that the "include" directive in this case is unnecessary because pyright includes all ".py" files in your project by default. Perhaps you intended to specify "extraPaths" instead? The "extraPaths" setting specifies extra import resolution paths. By default, pyright includes "src" in the "extraPaths" list because a directory of this name is commonly used for project sources.

In your initial bug report, it looks like you're running the language server rather than the CLI. The language server behavior will be impacted by any LSP settings that you've configured in your client. You haven't specified any LSP settings though, so I presume you're using the default settings.

**Comment by CamJN** [comment]
The venv is in the project dir, I do that to keep each project's venv isolated from other projects and the rest of the system. I just ignore the venv-related stuff with `.gitignore`.

I use the venv and venvPath approach because my editor doesn't inherit the env vars from a shell so I have to have pyright activate the venv for itself.

I am using the language server in my editor, however to simplify reproduction I gave instructions for the cli because then I don't have to tell you how to configure my editor, and the issue repros on the cil too, so that seemed an easier way for me to show you the problem. I do use the default settings for the lsp in my editor for what it's worth.

It's good to know i should have been using `extraPaths` for "src", I moved the "src" to `extraPaths` and that seems to have fixed it, which is strange if "src" is included in `extraPaths` by default, perhaps that's related to what changed?

In case it helps with repro:
```
sw_vers
ProductName:		macOS
ProductVersion:		15.0
BuildVersion:		24A335

python --version
Python 3.11.10
```

Both python3.11 and pyright are installed from homebrew.

Anyway, it's working for me now, and I suppose I was misusing the config before, so if you want to close this feel free, but it is weird that I had to put "src" in extraPaths if it's supposed to be there by default.

**Comment by erictraut** [comment]
@rchiodo, any thoughts on this? Is it possible that your recent changes in config merging affect the default value of "extraPaths"?

**Comment by erictraut** [comment]
> I use the venv and venvPath approach because my editor doesn't inherit the env vars from a shell so I have to have pyright activate the venv for itself.

In that case, the `python.pythonPath` config approach is preferable to "venvPath" and "venv". You can specify this in your LSP settings. The problem with the "venvPath" and "venv" approach is that pyright doesn't activate the python environment and instead simply attempts to guess as the import resolution paths that would be used if the venv were activated. With the `python.pythonPath` approach, pyright invokes the Python interpreter and uses a minimal script to dump its `sys.path` values. This tends to be much more robust. 

I've thought about removing support for "venvPath" and "venv" because that approach is unreliable. Unfortunately, there are some edge cases where it's still needed. I'll think about ways to make it clearer in the documentation that this approach should be used only if all other approaches do not work.
